### PR TITLE
expr: check timezone overflow

### LINKED ISSUE #20514: [SQLsmith] thread 'timely:work-0' panicked at '`NaiveDateTime + Duration` overflowed
### What version of Materialize are you using?
v0.61.0-dev (e4e0a0bac)

### What is the issue?

The query
```
 SELECT pg_catalog.timezone(
  'JAPAN',
  CAST((TIMESTAMPTZ '95143-12-31 23:59:59+06' + INTERVAL '167 MILLENNIUM') as timestamptz)
);
```
panics with
```
thread 'coordinator' panicked at '`NaiveDateTime + Duration` overflowed', /Users/nrainer/.cargo/registry/src/index.crates.io-6f17d22bba15001f/chrono-0.4.25/src/naive/datetime/mod.rs:1474:38
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
cluster-u1-replica-1: 2023-07-13T13:28:25.886568Z  WARN mz_persist_client::rpc: pubsub client error: Status { code: Unknown, message: "error reading a body from connection: stream closed because of a broken pipe", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
```

### Notes
* It is important to use `JAPAN` as time zone. It does not occur with a list of other time zones.
* This is not fixed by https://github.com/MaterializeInc/materialize/issues/20324 / https://github.com/MaterializeInc/materialize/pull/20386.

### Original SQLsmith query

For reference, the original query used by SQLsmith was

<details>
  <summary>this one</summary>

```
select  
  subq_0."c4" as c0
from 
  (select  
        ref_0."n_comment" as c0, 
        pg_catalog.int8range(
          CAST(ref_0."n_nationkey" as int8),
          CAST(ref_0."n_nationkey" as int8),
          CAST(ref_0."n_comment" as text)) as c1, 
        - (select pg_catalog.variance("id") from mz_internal.mz_records_per_dataflow_operator_per_worker)
             as c2, 
        ref_0."n_comment" as c3, 
        92 as c4, 
        (case when (ref_0."n_comment") ~~* (ref_0."n_name") then (select "e" from public.t limit 1 offset 2)
               else (select "e" from public.t limit 1 offset 2)
               end
            ) <= (case when (cast(null as int2)) = (cast(null as int2)) then cast(null as uint4) else cast(null as uint4) end
            ) as c5
      from 
        public.nation as ref_0
      where (pg_catalog.date(
          CAST(pg_catalog.timezone(
            CAST(ref_0."n_name" as text),
            CAST((TIMESTAMPTZ '95143-12-31 23:59:59+06' + INTERVAL '167 MILLENNIUM') as timestamptz)) as timestamp))) = ((select "o_orderdate" from public.orders limit 1 offset 1)
          )
      limit 131) as subq_0
where (subq_0."c5" is NULL) 
  or (((pg_catalog.jsonb_strip_nulls(
        CAST((select "details" from mz_internal.mz_sink_statuses limit 1 offset 1)
           as jsonb))) <@ ((subq_0."c3") || (subq_0."c3"))) 
    or (subq_0."c5" is not NULL))
limit 74;
```
</details>

using

<details>
  <summary>this setup</summary>

```
CREATE SOURCE tpch
              FROM LOAD GENERATOR TPCH (SCALE FACTOR 0.00001)
              FOR ALL TABLES
              WITH (SIZE = '1');
CREATE SOURCE auction
              FROM LOAD GENERATOR AUCTION (SCALE FACTOR 0.01)
              FOR ALL TABLES
              WITH (SIZE = '1');
CREATE SOURCE counter
              FROM LOAD GENERATOR COUNTER (SCALE FACTOR 0.0001)
              WITH (SIZE = '1');
CREATE TABLE t (a int2, b int4, c int8, d uint2, e uint4, f uint8, g text);
INSERT INTO t VALUES (1, 2, 3, 4, 5, 6, '7'), (3, 4, 5, 6, 7, 8, '9'), (5, 6, 7, 8, 9, 10, '11'), (7, 8, 9, 10, 11, 12, '13'), (9, 10, 11, 12, 13, 14, '15'), (11, 12, 13, 14, 15, 16, '17'), (13, 14, 15, 16, 17, 18, '19'), (15, 16, 17, 18, 19, 20, '21');
CREATE MATERIALIZED VIEW mv AS SELECT a + b AS col1, c + d AS col2, e + f AS col3, g AS col4 FROM t;
CREATE MATERIALIZED VIEW mv2 AS SELECT count(*) FROM mv;
CREATE DEFAULT INDEX ON mv;
```
</details>

failing with

<details>
  <summary>this backtrace</summary>

```
sqlsmith-mz_1-1  | thread 'timely:work-0' panicked at '`NaiveDateTime + Duration` overflowed', /cargo/registry/src/github.com-1ecc6299db9ec823/chrono-0.4.25/src/naive/datetime/mod.rs:1474:38
sqlsmith-mz_1-1  | stack backtrace:
sqlsmith-mz_1-1  |    0: rust_begin_unwind
sqlsmith-mz_1-1  |              at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/panicking.rs:575:5
sqlsmith-mz_1-1  |    1: core::panicking::panic_fmt
sqlsmith-mz_1-1  |              at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panicking.rs:64:14
sqlsmith-mz_1-1  |    2: core::panicking::panic_display
sqlsmith-mz_1-1  |              at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panicking.rs:147:5
sqlsmith-mz_1-1  |    3: core::panicking::panic_str
sqlsmith-mz_1-1  |              at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panicking.rs:131:5
sqlsmith-mz_1-1  |    4: core::option::expect_failed
sqlsmith-mz_1-1  |              at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/option.rs:1924:5
sqlsmith-mz_1-1  |    5: <core::option::Option<chrono::naive::datetime::NaiveDateTime>>::expect
sqlsmith-mz_1-1  |              at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/option.rs:786:21
sqlsmith-mz_1-1  |    6: <chrono::naive::datetime::NaiveDateTime as core::ops::arith::Add<chrono::oldtime::Duration>>::add
sqlsmith-mz_1-1  |              at ./cargo/registry/src/github.com-1ecc6299db9ec823/chrono-0.4.25/src/naive/datetime/mod.rs:1474:38
sqlsmith-mz_1-1  |    7: chrono::offset::fixed::add_with_leapsecond::<chrono::naive::datetime::NaiveDateTime>
sqlsmith-mz_1-1  |              at ./cargo/registry/src/github.com-1ecc6299db9ec823/chrono-0.4.25/src/offset/fixed.rs:188:5
sqlsmith-mz_1-1  |    8: <chrono::naive::datetime::NaiveDateTime as core::ops::arith::Add<chrono::offset::fixed::FixedOffset>>::add
sqlsmith-mz_1-1  |              at ./cargo/registry/src/github.com-1ecc6299db9ec823/chrono-0.4.25/src/offset/fixed.rs:214:9
sqlsmith-mz_1-1  |    9: mz_expr::scalar::func::impls::timestamp::timezone_timestamptz
sqlsmith-mz_1-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0ec6bb5e809cf2b5f-1/materialize/tests/src/expr/src/scalar/func/impls/timestamp.rs:471:5
sqlsmith-mz_1-1  |   10: <mz_expr::scalar::func::BinaryFunc>::eval::{closure#1}
sqlsmith-mz_1-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0ec6bb5e809cf2b5f-1/materialize/tests/src/expr/src/scalar/func.rs:2409:20
sqlsmith-mz_1-1  |   11: <core::result::Result<mz_repr::adt::datetime::Timezone, mz_expr::scalar::EvalError>>::and_then::<mz_repr::scalar::Datum, <mz_expr::scalar::func::BinaryFunc>::eval::{closure#1}>
sqlsmith-mz_1-1  |              at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/result.rs:1371:22
sqlsmith-mz_1-1  |   12: <mz_expr::scalar::func::BinaryFunc>::eval
sqlsmith-mz_1-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0ec6bb5e809cf2b5f-1/materialize/tests/src/expr/src/scalar/func.rs:2408:48
sqlsmith-mz_1-1  |   13: <mz_expr::scalar::MirScalarExpr>::eval
sqlsmith-mz_1-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0ec6bb5e809cf2b5f-1/materialize/tests/src/expr/src/scalar/mod.rs:1761:17
sqlsmith-mz_1-1  |   14: <mz_expr::scalar::func::impls::timestamp::CastTimestampToDate as mz_expr::scalar::func::LazyUnaryFunc>::eval
sqlsmith-mz_1-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0ec6bb5e809cf2b5f-1/materialize/tests/src/expr/src/scalar/func.rs:4237:41
sqlsmith-mz_1-1  |   15: <mz_expr::scalar::func::UnaryFunc>::eval
sqlsmith-mz_1-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0ec6bb5e809cf2b5f-1/materialize/tests/src/expr/src/scalar/func/macros.rs:343:41
sqlsmith-mz_1-1  |   16: <mz_expr::scalar::MirScalarExpr>::eval
sqlsmith-mz_1-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0ec6bb5e809cf2b5f-1/materialize/tests/src/expr/src/scalar/mod.rs:1759:56
sqlsmith-mz_1-1  |   17: <mz_compute::render::context::CollectionBundle<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::row::Row>>::ensure_collections::{closure#2}::{closure#0}
sqlsmith-mz_1-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0ec6bb5e809cf2b5f-1/materialize/tests/src/compute/src/render/context.rs:777:57
sqlsmith-mz_1-1  |   18: <&mut <mz_compute::render::context::CollectionBundle<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::row::Row>>::ensure_collections::{closure#2}::{closure#0} as core::ops::function::FnOnce<(&mz_expr::scalar::MirScalarExpr,)>>::call_once
sqlsmith-mz_1-1  |              at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/ops/function.rs:310:13
sqlsmith-mz_1-1  |   19: <core::option::Option<&mz_expr::scalar::MirScalarExpr>>::map::<core::result::Result<mz_repr::scalar::Datum, mz_expr::scalar::EvalError>, &mut <mz_compute::render::context::CollectionBundle<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::Timestamp>, mz_repr::row::Row>>::ensure_collections::{closure#2}::{closure#0}>
```
</details>

### Related
* https://github.com/MaterializeInc/materialize/issues/20324

```
ci-regexp: thread 'coordinator' panicked at '`NaiveDateTime + Duration` overflowed'
```

### PR DESCRIPTION
**Pull request by mjibson**
Needed to add our own checked method because chrono doesn't have one for this specific type pair.

Add some additional interesting datums that would have caught this.

Fixes #20514

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

File: src/expr/src/scalar/func/impls/timestamp.rs [hunk 1]
```diff
@@ -463,12 +465,22 @@ pub fn timezone_timestamp(
...
     };
-    utc.naive_utc() + offset
+    checked_add_with_leapsecond(&utc.naive_utc(), &offset).ok_or(EvalError::TimestampOutOfRange)
+}
+
+/// Checked addition that is missing from chrono. Adapt its methods here but add a check.
+fn checked_add_with_leapsecond(lhs: &NaiveDateTime, rhs: &FixedOffset) -> Option<NaiveDateTime> {
+    // extract and temporarily remove the fractional part and later recover it
+    let nanos = lhs.nanosecond();
+    let lhs = lhs.with_nanosecond(0).unwrap();
```
Comment (guswynn) [inline src/expr/src/scalar/func/impls/timestamp.rs:480]: might as well `?` here?

**Review by guswynn — APPROVED** [review]
seemsreasonable

File: src/expr/src/scalar/func/impls/timestamp.rs (hunk 1 above)
Comment (mjibson) [reply src/expr/src/scalar/func/impls/timestamp.rs:480]: I think it's better as is because we should never be returning None here.

**Comment by nrainer-materialize** [comment]
Thanks!

**Pull request merged**